flax/experimental/nnx/nnx/graph_utils.py

```diff
@@ -455,16 +459,19 @@ def _graph_flatten(
     if is_node(value):
       graphdef = _graph_flatten((*path, key), ref_to_index, flat_state, value)
       subgraphs.append((key, graphdef))
-    elif isinstance(value, Variable):
+    elif is_state_leaf(value):
       str_path = '/'.join((*path, key))
-      if value in ref_to_index:
-        variables.append((key, ref_to_index[value]))
+      if isinstance(value, Variable):
```

Comment:
Nit: maybe move this into a separate branch before `is_state_leaf` to reduce the nesting here?

```
elif isinstance(value, Variable):
  ...
elif is_state_leaf(value):
  ...
```